cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskTracker.java

```diff
@@ -4,51 +4,149 @@
+  private static final String DEAD = "dead";
+  private static final String COMPLETED = "completed";
+  private static final String GAUGE_ONGOING_EXECUTION_IN_KAFKA_ASSIGNER_MODE = "ongoing-execution-kafka_assigner";
+  private static final String GAUGE_ONGOING_EXECUTION_IN_NON_KAFKA_ASSIGNER_MODE = "ongoing-execution-non_kafka_assigner";
 
-    for (ExecutionTask.State state : states) {
-      _replicaActionTasks.put(state, new HashSet<>());
-      _leaderActionTasks.put(state, new HashSet<>());
+  ExecutionTaskTracker(MetricRegistry dropwizardMetricRegistry, Time time) {
+    List<State> states = State.cachedValues();
+    List<TaskType> taskTypes = TaskType.cachedValues();
+    _tasksByType = new HashMap<>(taskTypes.size());
+    for (TaskType type : taskTypes) {
+      Map<State, Set<ExecutionTask>> taskMap = new HashMap<>(states.size());
+      for (State state : states) {
+        taskMap.put(state, new HashSet<>());
+      }
+      _tasksByType.put(type, taskMap);
     }
+    _remainingDataToMoveInMB = 0L;
+    _inExecutionDataMovementInMB = 0L;
+    _finishedDataMovementInMB = 0L;
     _isKafkaAssignerMode = false;
+    _time = time;
+
+    // Register gauge sensors.
+    registerGaugeSensors(dropwizardMetricRegistry);
+  }
+
+  private void registerGaugeSensors(MetricRegistry dropwizardMetricRegistry) {
+    String metricName = "Executor";
+    for (TaskType type : TaskType.cachedValues()) {
+      for (State state : State.cachedValues()) {
+        String typeString =  type == TaskType.REPLICA_ACTION ? REPLICA_ACTION :
+                                                               LEADERSHIP_ACTION;
+        String stateString =  state == State.PENDING     ? PENDING :
+                              state == State.IN_PROGRESS ? IN_PROGRESS :
+                              state == State.ABORTING    ? ABORTING :
+                              state == State.ABORTED     ? ABORTED :
+                              state == State.COMPLETED   ? COMPLETED :
+                                                           DEAD;
+        dropwizardMetricRegistry.register(MetricRegistry.name(metricName, typeString + "-" + stateString),
+                                          (Gauge<Integer>) () -> _tasksByType.get(type).get(state).size());
+      }
+    }
+    dropwizardMetricRegistry.register(MetricRegistry.name(metricName, GAUGE_ONGOING_EXECUTION_IN_KAFKA_ASSIGNER_MODE),
+                                      (Gauge<Integer>) () -> _isKafkaAssignerMode
+                                                             && !inExecutionTasks(TaskType.cachedValues()).isEmpty() ? 1 : 0);
+    dropwizardMetricRegistry.register(MetricRegistry.name(metricName, GAUGE_ONGOING_EXECUTION_IN_NON_KAFKA_ASSIGNER_MODE),
+                                      (Gauge<Integer>) () -> !_isKafkaAssignerMode
+                                                             && !inExecutionTasks(TaskType.cachedValues()).isEmpty() ? 1 : 0);
```

Comment:
Did we verify (e.g. via `jConsole`) that the existing metrics are not renamed/removed after this patch?

Reply:
Yes, I checked on a local deployment.